File: crawler/security/dns_policy.py
"""Injectable DNS resolver and all-address validation for TASK-022B."""

from __future__ import annotations

import ipaddress
from typing import List, Protocol, Sequence

from crawler.security.ip_policy import classify_ip
from crawler.security.models import DNSValidationResult


class Resolver(Protocol):
    def resolve(self, host: str, timeout_seconds: float) -> List[str]:
        """Return all addresses for an ASCII hostname or raise."""
        ...


class DNSValidationError(ValueError):
    def __init__(self, reason_code: str, message: str = "") -> None:
        super().__init__(message or reason_code)
        self.reason_code = reason_code


def _stable_sort(addresses: Sequence[str]) -> List[str]:
    parsed = []
    seen = set()
    for address in addresses:
        try:
            ip = ipaddress.ip_address(address)
        except ValueError:
            raise DNSValidationError("invalid_dns_answer")
        canonical = str(ip)
        if canonical in seen:
            continue
        seen.add(canonical)
        parsed.append(ip)
    parsed.sort(key=lambda ip: (ip.version, ip.packed))
    return [str(ip) for ip in parsed]


def resolve_and_validate(
    host: str,
    resolver: Resolver,
    *,
    timeout_seconds: float = 5.0,
    max_addresses: int = 16,
) -> DNSValidationResult:
    if not host:
        raise DNSValidationError("invalid_host")

    try:
        ipaddress.ip_address(host)
        raw_addresses = [host]
    except ValueError:
        try:
            raw_addresses = list(resolver.resolve(host, timeout_seconds))
        except TimeoutError:
            raise DNSValidationError("dns_timeout")
        except Exception:
            raise DNSValidationError("dns_failed")

    addresses = _stable_sort(raw_addresses)
    if not addresses:
        raise DNSValidationError("dns_no_addresses")
    if len(addresses) > max_addresses:
        raise DNSValidationError("dns_too_many_addresses")

    for address in addresses:
        classification = classify_ip(address)
        if not classification.allowed:
            return DNSValidationResult(
                host=host,
                addresses=tuple(addresses),
                allowed=False,
                reason_code="ip_not_allowed",
            )

    return DNSValidationResult(
        host=host,
        addresses=tuple(addresses),
        allowed=True,
        reason_code="",
    )

File: go-spider/internal/security/dns_policy.go
package security

import (
	"bytes"
	"context"
	"errors"
	"net/netip"
	"sort"
)

type Resolver interface {
	Resolve(ctx context.Context, host string) ([]string, error)
}

type DNSValidationError struct {
	ReasonCode string
}

func (e *DNSValidationError) Error() string { return e.ReasonCode }

type sortedAddr struct {
	canonical string
	version   int
	key       []byte
}

func sortAddresses(addresses []string) ([]string, error) {
	seen := make(map[string]struct{}, len(addresses))
	items := make([]sortedAddr, 0, len(addresses))
	for _, raw := range addresses {
		addr, err := netip.ParseAddr(raw)
		if err != nil {
			return nil, &DNSValidationError{ReasonCode: "invalid_dns_answer"}
		}
		canonical := addr.String()
		if _, ok := seen[canonical]; ok {
			continue
		}
		seen[canonical] = struct{}{}
		version := 4
		var key []byte
		if addr.Is4() {
			v4 := addr.As4()
			key = v4[:]
		} else {
			version = 6
			v6 := addr.As16()
			key = v6[:]
		}
		items = append(items, sortedAddr{canonical: canonical, version: version, key: key})
	}
	sort.SliceStable(items, func(i, j int) bool {
		if items[i].version != items[j].version {
			return items[i].version < items[j].version
		}
		return bytes.Compare(items[i].key, items[j].key) < 0
	})
	out := make([]string, 0, len(items))
	for _, item := range items {
		out = append(out, item.canonical)
	}
	return out, nil
}

func ResolveAndValidate(ctx context.Context, host string, resolver Resolver, maxAddresses int) (DNSValidationResult, error) {
	if host == "" {
		return DNSValidationResult{}, &DNSValidationError{ReasonCode: "invalid_host"}
	}
	var raw []string
	if _, err := netip.ParseAddr(host); err == nil {
		raw = []string{host}
	} else {
		if resolver == nil {
			return DNSValidationResult{}, &DNSValidationError{ReasonCode: "dns_failed"}
		}
		addresses, err := resolver.Resolve(ctx, host)
		if err != nil {
			if errors.Is(err, context.DeadlineExceeded) || errors.Is(ctx.Err(), context.DeadlineExceeded) {
				return DNSValidationResult{}, &DNSValidationError{ReasonCode: "dns_timeout"}
			}
			return DNSValidationResult{}, &DNSValidationError{ReasonCode: "dns_failed"}
		}
		raw = addresses
	}

	addresses, err := sortAddresses(raw)
	if err != nil {
		return DNSValidationResult{}, err
	}
	if len(addresses) == 0 {
		return DNSValidationResult{}, &DNSValidationError{ReasonCode: "dns_no_addresses"}
	}
	if len(addresses) > maxAddresses {
		return DNSValidationResult{}, &DNSValidationError{ReasonCode: "dns_too_many_addresses"}
	}
	for _, address := range addresses {
		classification, err := ClassifyIP(address)
		if err != nil {
			return DNSValidationResult{}, &DNSValidationError{ReasonCode: "invalid_dns_answer"}
		}
		if !classification.Allowed {
			return DNSValidationResult{
				Host:       host,
				Addresses:  addresses,
				Allowed:    false,
				ReasonCode: "ip_not_allowed",
			}, nil
		}
	}
	return DNSValidationResult{
		Host:      host,
		Addresses: addresses,
		Allowed:   true,
	}, nil
}
